enhancements/installer/aws-custom-region-and-endpoints.md

```diff
@@ -0,0 +1,402 @@
+
+2. The installer should ensure that the service endpoint is at least reachable from the host.
+
+3. The URL for the service endpoint must be `https` and the host should trust the certificate.
+
+#### Configuring the AWS SDK
+
+There are 2 ways to configure the AWS SDK,
+
+1. Use the sdk [resolver][aws-sdk-endpoints-resolver] like Kubernetes AWS [cloud-provider][k8s-aws-cloud-provider-get-resolver] so that the AWS SDK's default resolver is used for AWS APIs and override with the user provided ones when available.
+
+2. Specifically override the AWS config when initializing the clients, for example the terraform-provider-aws's internal [configuration][terraform-provider-aws-config-endpoints]
+
+#### Configuring the internal terraform-provider-aws with service overrides
+
+The Terraform AWS provider doesn't accept random AWS API endpoint [overrides][terraform-aws-provider-api-endpoints], rather endpoint for each service needs to be explicitly specified. So the `tfvars` can pass a map object and the terraform `main.tf` can set each required endpoint override when certain key is present and set in the provide map object.
+
+#### Destroying clusters
+
+The `metadata.json` needs to store the custom endpoints from the `install-config.yaml` so that the users can delete the clusters without any previous state.
+
+### Infrastructure global configuration
+
+Since almost of the cluster operators that communicate to the the AWS API need the use the API endpoints provided by the user, the `Infrastructure` global configuration seems like the best place to store and discover this information.
+
+And in contrast to the current configuration in the `Infrastructure` object like `region` is not editable as day-2 operation, the users should be able to edit the AWS endpoints day-2, therefore this configuration best belongs in the `spec` section.
+
+```go
+// InfrastructureSpec contains settings that apply to the cluster infrastructure.
+type InfrastructureSpec struct {
+  // platformSpec holds configuration specific to the underlying
+  // infrastructure provider.
+  // +optional
+  PlatformSpec *PlatformSpec `json:"platformSpec,omitempty"`
+}
+
+// PlatformSpec holds some configuration to the underlying infrastructure provider
+// of the current cluster. It is supposed that only one of the spec structs is set.
+type PlatformSpec struct {
+  // AWS contains settings specific to the Amazon Web Services infrastructure provider.
+  // +optional
+  AWS *AWSPlatformSpec `json:"aws,omitempty"`
+
+  ...
+}
+```
+
+```go
+// AWSPlatformSpec holds the current status of the Amazon Web Services infrastructure provider.
+type AWSPlatformSpec struct {
+  // ServiceEndpoints list contains custom endpoints which will override default
+  // service endpoint of AWS Services.
+  // There must be only one ServiceEndpoint for a service.
+  // +optional
+  ServiceEndpoints []AWSServiceEndpoint `json:"serviceEndpoints,omitempty"`
```

Comment:
Shouldn't this be a map-based type? I also recommend that we have a string-based type to hold the constants for AWS services that need overrides.

```
type AWSService string

const (
    IAM AWSService = "iam"
    EC2 AWSService = "ec2"
    ELB AWSService = "elb"
    Route53 AWSService = "route53"
    ...
)
```

Reply:
> I also recommend that we have a string-based type to hold the constants for AWS services that need overrides.

maintaining a separate list is too much overhead, and there are _infinite_ AWS services. if OLM based operators start using this service endpoints list.. we wil be stuck adding types for those operators.

I think we can run validations esp when we move from spec to status to see if the service name is unknown.. that too is going to be tough unless aws sdk has a list ready :/

> Shouldn't this be a map-based type?

I find the list format to be more ergonomic... but i think we can use maps vs list based on

which one is easy to show validation error and which one might be easy for patching/updating.

```
ec2:
  url: ...
s3:
  url: ...
ec2:
  url: ...
```
I'm not sure if the parser itself will error out or just use the later key value...